### ISSUE TITLE
Bug in lib/Net/Domain.pm

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6779]
Migrated from [rt.perl.org#23953](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=23953) (status was 'resolved')

Searchable as RT23953$

**Comment by p5pRT** [comment]
### From frankj@osc.no

#### Created by frankj@osc.no


This is a bug report for perl from frankj@&#8203;osc\.no\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
There is a bug in lib/Net/Domain\.pm which can cause any perl program
to use up all memory\.  The conditions for this is however rather
obscure&#8203;:  DNS on the host will have to be disabled\, NIS enabled\,
and the NIS domainname must begin with "\." \(Or contain two "\."'s in
a row\)\.

This bug probably exists in \_all\_ versions of perl\, I have verified
its existance in version 5\.6\.1\, 5\.8 and 5\.8\.1\-RC4\.

Here is a patch&#8203;:
===================================================================
diff \-cr perl\-5\.8\.1\-RC4/lib/Net/Domain\.pm perl\-5\.8\.1\-RC4\-fix/lib/Net/Domain\.pm
\*\*\* perl\-5\.8\.1\-RC4/lib/Net/Domain\.pm    Thu Jul 10 08&#8203;:11&#8203;:51 2003
\-\-\- perl\-5\.8\.1\-RC4\-fix/lib/Net/Domain\.pm        Mon Sep 22 13&#8203;:44&#8203;:20 2003
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 176\,181 \*\*\*\*
\-\-\- 176\,182 \-\-\-\-
&nbsp;            my @&#8203;h = \(\);
&nbsp;            while\(length\($dom\)\) \{
&nbsp;                push\(@&#8203;h\, "$host\.$dom"\);
\+               $dom =~ s/^\\\.//g;
&nbsp;                $dom =~ s/^\[^\.\]\+\.//;
&nbsp;            \}
&nbsp;            unshift\(@&#8203;hosts\,@&#8203;h\);
===================================================================
As you can see from the original code above\, if $dom contains an extranous
"\."\, the loop will never finish\.  It will also gobble up all memory due
to the push command\.
The fix simply removes such extra "\."'s so that if domainname is set
to "\.bar\.com"\, the loop above will first find "bar\.com" and then "com"\.
There are of course many other ways this loop could be fixed\, but this
fix has been verified to work\.

We encountered this bug at a customer of ours\, where they had chosen to
have NIS domainname to "\.xxx\.yy\.co\.jp"\.  And they said that according
to Solaris specification\, NIS domain allows this syntax\.  Unfortunatedly
Domain\.pm just assumes no extra "\."\, since DNS doesn't allow it\.

\-Frank\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.0:

Configured by bhcompile at Tue Feb 18 22:17:47 EST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.20-2.48smp, archname=i386-linux-thread-multi
    uname='linux stripples.devel.redhat.com 2.4.20-2.48smp #1 smp thu feb 13 11:44:55 est 2003 i686 i686 i386 gnulinux '
    config_args='-des -Doptimize=-O2 -march=i386 -mcpu=i686 -g -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Darchname=i386-linux -Dvendorprefix=/usr -Dsiteprefix=/usr -Dotherlibdirs=/usr/lib/perl5/5.8.0 -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -march=i386 -mcpu=i686 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.2 20030213 (Red Hat Linux 8.0 3.2.2-1)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.3.1.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.8.0/i386-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    MAINT18379


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.0
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/5.8.0
    .


Environment for perl v5.8.0:
    HOME=/u/pompel/b/frankj
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/local/psm_dev/v5/bin:/local/psm_dev/v4/bin:/local/psm_dev/v3/bin:/usr/bsd:/local/gnu/bin:/sbin:/usr/sbin:/u/serum/e1/bin/sun4:/usr/ucb:/usr/local/games/bin:/u/serum/e1/frankj/bin:/usr/local/bin:/usr/ucb:/bin:/usr/bin:/etc:/usr/etc:/local/mtools/bin/:/local/common/bin:/local/X11R5/bin:/local/bin:/local/gnu/bin:/local/osacc/bin:/local/osacc/etc:/usr/osacc/bin:/usr/osacc/etc:/usr/games:/usr/bin/X11:/usr/openwin/bin:.
    PERL_BADLANG (unset)
    SHELL=/bin/sh

```
</details>

**Comment by p5pRT** [comment]
### From @tamias

On Mon\, Sep 22\, 2003 at 11&#8203;:57&#8203;:53AM \-0000\, frankj@&#8203;osc\.no \(via RT\) wrote&#8203;:

> diff \-cr perl\-5\.8\.1\-RC4/lib/Net/Domain\.pm perl\-5\.8\.1\-RC4\-fix/lib/Net/Domain\.pm
> \*\*\* perl\-5\.8\.1\-RC4/lib/Net/Domain\.pm    Thu Jul 10 08&#8203;:11&#8203;:51 2003
> \-\-\- perl\-5\.8\.1\-RC4\-fix/lib/Net/Domain\.pm        Mon Sep 22 13&#8203;:44&#8203;:20 2003
> \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
> \*\*\* 176\,181 \*\*\*\*
> \-\-\- 176\,182 \-\-\-\-
>             my @&#8203;h = \(\);
>             while\(length\($dom\)\) \{
>                 push\(@&#8203;h\, "$host\.$dom"\);
> \+               $dom =~ s/^\\\.//g;
>                 $dom =~ s/^\[^\.\]\+\.//;
>             \}
>             unshift\(@&#8203;hosts\,@&#8203;h\);
> ===================================================================
> As you can see from the original code above\, if $dom contains an extranous
> "\."\, the loop will never finish\.  It will also gobble up all memory due
> to the push command\.

I believe that code will also enter an infinite loop if $dom ends up being
one character long\.  \(Presumably this shouldn't happen with well\-formed
domains\, but the potential for an infinite loop should still be corrected\.\)

The easiest fix is to change the second substitution&#8203;:

$dom =~ s/^\[^\.\]\+\.?//;

A better fix might be to test specifically for $dom being one character
long and returning some sort of error\.  I'd say that's for someone who
knows more about the module than I to determine\.

Ronald

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @Hugmeir

On Mon Sep 22 11&#8203;:39&#8203;:01 2003\, rjk\-perl\-p5p@&#8203;tamias\.net wrote&#8203;:
> On Mon\, Sep 22\, 2003 at 11&#8203;:57&#8203;:53AM \-0000\, frankj@&#8203;osc\.no \(via RT\)
> wrote&#8203;:
> 
> > diff \-cr perl\-5\.8\.1\-RC4/lib/Net/Domain\.pm perl\-5\.8\.1\-RC4\-
> fix/lib/Net/Domain\.pm
> > \*\*\* perl\-5\.8\.1\-RC4/lib/Net/Domain\.pm    Thu Jul 10 08&#8203;:11&#8203;:51 2003
> > \-\-\- perl\-5\.8\.1\-RC4\-fix/lib/Net/Domain\.pm        Mon Sep 22 13&#8203;:44&#8203;:20
> 2003
> > \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
> > \*\*\* 176\,181 \*\*\*\*
> > \-\-\- 176\,182 \-\-\-\-
> >             my @&#8203;h = \(\);
> >             while\(length\($dom\)\) \{
> >                 push\(@&#8203;h\, "$host\.$dom"\);
> > \+               $dom =~ s/^\\\.//g;
> >                 $dom =~ s/^\[^\.\]\+\.//;
> >             \}
> >             unshift\(@&#8203;hosts\,@&#8203;h\);
> > ===================================================================
> > As you can see from the original code above\, if $dom contains an
> extranous
> > "\."\, the loop will never finish\.  It will also gobble up all memory
> due
> > to the push command\.
> 
> I believe that code will also enter an infinite loop if $dom ends up
> being
> one character long\.  \(Presumably this shouldn't happen with well\-
> formed
> domains\, but the potential for an infinite loop should still be
> corrected\.\)
> 
> The easiest fix is to change the second substitution&#8203;:
> 
> $dom =~ s/^\[^\.\]\+\.?//;
> 
> A better fix might be to test specifically for $dom being one
> character
> long and returning some sort of error\.  I'd say that's for someone who
> knows more about the module than I to determine\.
> 
> Ronald
> 


Looks like this was already applied to Net&#8203;::Domain\. This ticket was
closed for a while in 2008\, then reopened by stmpeters\. I'm guessing
that was a mistake? Or is there any reason why this shouldn't be marked
as resolved again?

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Apr 28 03&#8203;:16&#8203;:11 2012\, Hugmeir wrote&#8203;:
> On Mon Sep 22 11&#8203;:39&#8203;:01 2003\, rjk\-perl\-p5p@&#8203;tamias\.net wrote&#8203;:
> > On Mon\, Sep 22\, 2003 at 11&#8203;:57&#8203;:53AM \-0000\, frankj@&#8203;osc\.no \(via RT\)
> > wrote&#8203;:
> > 
> > > diff \-cr perl\-5\.8\.1\-RC4/lib/Net/Domain\.pm perl\-5\.8\.1\-RC4\-
> > fix/lib/Net/Domain\.pm
> > > \*\*\* perl\-5\.8\.1\-RC4/lib/Net/Domain\.pm    Thu Jul 10 08&#8203;:11&#8203;:51 2003
> > > \-\-\- perl\-5\.8\.1\-RC4\-fix/lib/Net/Domain\.pm        Mon Sep 22 13&#8203;:44&#8203;:20
> > 2003
> > > \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
> > > \*\*\* 176\,181 \*\*\*\*
> > > \-\-\- 176\,182 \-\-\-\-
> > >             my @&#8203;h = \(\);
> > >             while\(length\($dom\)\) \{
> > >                 push\(@&#8203;h\, "$host\.$dom"\);
> > > \+               $dom =~ s/^\\\.//g;
> > >                 $dom =~ s/^\[^\.\]\+\.//;
> > >             \}
> > >             unshift\(@&#8203;hosts\,@&#8203;h\);
> > > ===================================================================
> > > As you can see from the original code above\, if $dom contains an
> > extranous
> > > "\."\, the loop will never finish\.  It will also gobble up all memory
> > due
> > > to the push command\.
> > 
> > I believe that code will also enter an infinite loop if $dom ends up
> > being
> > one character long\.  \(Presumably this shouldn't happen with well\-
> > formed
> > domains\, but the potential for an infinite loop should still be
> > corrected\.\)
> > 
> > The easiest fix is to change the second substitution&#8203;:
> > 
> > $dom =~ s/^\[^\.\]\+\.?//;
> > 
> > A better fix might be to test specifically for $dom being one
> > character
> > long and returning some sort of error\.  I'd say that's for someone who
> > knows more about the module than I to determine\.
> > 
> > Ronald
> > 
> 
> 
> Looks like this was already applied to Net&#8203;::Domain\. This ticket was
> closed for a while in 2008\, then reopened by stmpeters\. I'm guessing
> that was a mistake? Or is there any reason why this shouldn't be marked
> as resolved again?

I know I’m not answering your question\, but this is in the cpan/
directory in blead\.  Porting/Maintainers\.pl lists upstream as undef\, but
I know that Graham Barr’s repository \(I don’t remember where that is\)
has patches applied that are not in the CPAN release\, making CPAN
effectively upstream\, I suppose\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@Hugmeir - Status changed from 'open' to 'resolved'
